packages/sanity/src/desk/comments/src/hooks/use-mention-options/useMentionOptions.ts

```diff
@@ -0,0 +1,118 @@
+/* eslint-disable max-nested-callbacks */
+import {useState, useEffect, useMemo} from 'react'
+import {Observable, concat, forkJoin, map, mergeMap, of, switchMap} from 'rxjs'
+import {SanityDocument} from '@sanity/client'
+import {sortBy} from 'lodash'
+import {Loadable, MentionOptionUser, MentionOptionsHookValue} from '../../types'
+import {grantsPermissionOn} from './helpers'
+import {
+  useProjectStore,
+  useUserStore,
+  useClient,
+  DEFAULT_STUDIO_CLIENT_OPTIONS,
+  ProjectData,
+} from 'sanity'
+
+const INITIAL_STATE: MentionOptionsHookValue = {
+  data: [],
+  error: null,
+  loading: true,
+}
+
+export interface MentionHookOptions {
+  documentValue: SanityDocument | null
+}
+
+let cachedSystemGroups: [] | null = null
+
+export function useMentionOptions(opts: MentionHookOptions): MentionOptionsHookValue {
+  const {documentValue} = opts
+
+  const projectStore = useProjectStore()
+  const userStore = useUserStore()
+  const client = useClient(DEFAULT_STUDIO_CLIENT_OPTIONS)
+
+  const [state, setState] = useState<MentionOptionsHookValue>(INITIAL_STATE)
+
+  const list$ = useMemo(() => {
+    // 1. Get the project members and filter out the robot users
+    const members$: Observable<ProjectData['members']> = projectStore
+      .get()
+      .pipe(map((res: ProjectData) => res.members?.filter((m) => !m.isRobot)))
+
+    // 2. Map the members to users to get more data of the users such as displayName (used for filtering)
+    const users$: Observable<MentionOptionUser[]> = members$.pipe(
+      switchMap(async (members) => {
+        const ids = members.map(({id}) => id)
+        const users = await userStore.getUsers(ids)
+        return users
+      }),
+      map((res) =>
+        res.map((user) => ({
+          displayName: user.displayName,
+          id: user.id,
+          canBeMentioned: false,
+        })),
+      ),
+    )
+
+    // 3. Get all the system groups. Use the cached response if it exists to avoid unnecessary requests.
+    const cached = cachedSystemGroups
+    const systemGroup$ = cached ? of(cached) : client.observable.fetch('*[_type == "system.group"]')
+
+    // 4. Check if the user has read permission on the document and set the `canBeMentioned` property
+    const grants$: Observable<MentionOptionUser[]> = forkJoin([users$, systemGroup$]).pipe(
```

Comment:
As `forkJoin` will wait for both streams to complete,  this stream will only emit a single (tuple) value and then complete. Just making sure this is the desired behavior, as it means a user will have to reload the studio to be able to mention a newly added user (e.g. a user that was added to the project after this ran initially) – in the future we may want to ma ke it reactive?

Reply:
This is the intended behavior and I think it's fine – what do you think? It would be nice to make it reactive in the future – but I don't think we need to worry about it right now!